src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs

```diff
@@ -411,13 +412,42 @@ private static object CreateInstance([DynamicallyAccessedMembers(DynamicallyAcce
                 throw new InvalidOperationException(SR.Format(SR.Error_CannotActivateAbstractOrInterface, type));
             }
 
-            if (!type.IsValueType)
+            ConstructorInfo[] constructors = type.GetConstructors(DeclaredOnlyLookup);
+
+            bool hasParameterlessPublicConstructor =
+                constructors.Any(ctor => ctor.IsPublic && ctor.GetParameters().Length == 0);
+
+            if (constructors.Length > 0 && !hasParameterlessPublicConstructor)
             {
-                bool hasDefaultConstructor = type.GetConstructors(DeclaredOnlyLookup).Any(ctor => ctor.IsPublic && ctor.GetParameters().Length == 0);
-                if (!hasDefaultConstructor)
+                // if the only constructor is private, then throw
+                if (constructors.Length == 1 && !constructors[0].IsPublic)
                 {
                     throw new InvalidOperationException(SR.Format(SR.Error_MissingParameterlessConstructor, type));
                 }
+
+                // find the biggest constructor so that we can bind to the most parameters
+                ParameterInfo[] parameters = constructors[0].GetParameters();
```

Comment:
Seems like this part needs work on how to find the biggest constructor.

Reply:
Yes. I changed this to discard ref and out parameters, and it now doesn't default to the first constructor. If no suitable constructor is found, it falls through to the old way (`Activator.CreateInstance`) which will, if there really isn't a suitable constructor, throw a 'no parameterless constructor' exception.